ml-prims/src/selection/columnWiseSort.h

```diff
@@ -0,0 +1,310 @@
+ */
+
+#pragma once
+
+#include <cuda_runtime.h>
+#include <cstddef>
+#include "cuda_utils.h"
+
+#include <cub/cub.cuh>
+#include <limits>
+
+#define ALIGN_BYTE 256
+#define ALIGN_MEMORY(x) (x + ALIGN_BYTE - 1) & ~(ALIGN_BYTE - 1)
+
+#define VALIDATE_TEMPLATE (is_same< InType, float >::value && BLOCK_SIZE <= 512) || \
+    (is_same< InType, double >::value && \
+    (is_same< OutType, int >::value || is_same< OutType, float >::value) &&       \
+           BLOCK_SIZE <= 512 && ITEMS_PER_THREAD <= 8) || \
+    (is_same< InType, double >::value && is_same< OutType, double >::value &&    \
+          BLOCK_SIZE <= 512 && ITEMS_PER_THREAD <= 6)
+
+#define INST_BLOCK_SORT(keyIn, keyOut, valueInOut, rows, columns, blockSize, elemPT, stream) \
+devKeyValSortColumnPerRow<InType, OutType, blockSize, elemPT><<<rows, blockSize, 0, stream>>> \
+        (keyIn, keyOut, valueInOut, rows, columns, numeric_limits<InType>::max())
+
+namespace MLCommon {
+namespace Selection {
+
+using namespace std;
+
+template <typename InType>
+__global__ void devLayoutIdx(InType *in, int n_cols, int totalElements) {
+  int idx = threadIdx.x + blockDim.x * blockIdx.x;
+  int n = n_cols;
+
+  if (idx < totalElements) {
+    in[idx] = idx % n;
+  }
+}
+
+template <typename T>
+__global__ void devOffsetKernel(T *in, T value, int n_times) {
+  int idx = threadIdx.x + blockIdx.x * blockDim.x;
+  if (idx < n_times)
+    in[idx] = idx * value;
+}
+
+// block level radix sort - can only sort as much data we can fit within shared memory
+template <
+    typename    InType,
+    typename    OutType,
+    int         BLOCK_SIZE,
+    int         ITEMS_PER_THREAD,
+    typename    std::enable_if<VALIDATE_TEMPLATE,  InType>::type* = nullptr>
+__global__ void devKeyValSortColumnPerRow(const InType *inputKeys, InType *outputKeys, 
+                                            OutType *inputVals, int n_rows, 
+                                            int n_cols, InType MAX_VALUE) {
+
+  typedef cub::BlockLoad<InType, BLOCK_SIZE, ITEMS_PER_THREAD, 
+                          cub::BLOCK_LOAD_WARP_TRANSPOSE> BlockLoadTypeKey;
+  typedef cub::BlockLoad<OutType, BLOCK_SIZE, ITEMS_PER_THREAD, 
+                          cub::BLOCK_LOAD_WARP_TRANSPOSE> BlockLoadTypeVal;
+
+  typedef cub::BlockRadixSort<InType, BLOCK_SIZE, ITEMS_PER_THREAD, 
+                                OutType> BlockRadixSortType;
+
+  __shared__ union TempStorage
+  {
+    struct {
+      typename BlockLoadTypeKey::TempStorage keyLoad;
+      typename BlockLoadTypeVal::TempStorage valLoad;
+    } load;
+    typename BlockRadixSortType::TempStorage sort;
+  } tempStorage;
+
+  InType threadKeys[ITEMS_PER_THREAD];
+  OutType threadValues[ITEMS_PER_THREAD];
+
+  int blockOffset = blockIdx.x * n_cols;
+  BlockLoadTypeKey(tempStorage.load.keyLoad).Load(inputKeys + blockOffset, threadKeys, n_cols, MAX_VALUE);
+  BlockLoadTypeVal(tempStorage.load.valLoad).Load(inputVals + blockOffset, threadValues, n_cols, MAX_VALUE);
```

Comment:
This extra load, as well as the extra store in the `devLayoutIdx` kernel can be avoided if the logic in that kernel can be fused here itself, right?

I agree that this is a perf feature. And as such, if we need this prim for 0.7 release, I'm ok if we file an issue for this, so that we don't miss doing this.

Reply:
This prim is not needed for 0.7

Reply:
I see. Thanks Corey.

@chirayuG-nvidia I think then it is better to implement this "fusing" technique along with this PR itself. Let me know if you think otherwise.

Reply:
@teju85 I have fused the devlayoutIdx kernel in this kernel. This helps save shared memory usage per block as well. We can know sort more data per-block but its not trivial to calculate upper-limit on that anymore. 

```
__shared__ union TempStorage
  {
    struct {
      typename BlockLoadTypeKey::TempStorage keyLoad;
    } load;
    typename BlockRadixSortType::TempStorage sort;
  } tempStorage;
```
Earlier size of TempStorage was dominated by key-value pair, now I am not able to figure out how to compute the smem per block to put guards for kernel instantiations.

Reply:
Good catch. Didn't realize that it'd also save smem usage!

Maybe we could do a similar struct thingy to compute smem usage? For example:
```cpp
template <typename InType, int BLOCK_SIZE, int ITEMS_PER_THREAD>
struct SmemBlock {
  typedef cub::BlockLoad<InType, BLOCK_SIZE, ITEMS_PER_THREAD, 
                          cub::BLOCK_LOAD_WARP_TRANSPOSE> BlockLoadTypeKey;
  typedef cub::BlockRadixSort<InType, BLOCK_SIZE, ITEMS_PER_THREAD, 
                                OutType> BlockRadixSortType;

  union TempStorage {
    struct {
      typename BlockLoadTypeKey::TempStorage keyLoad;
    } load;
    typename BlockRadixSortType::TempStorage sort;
  };
};
```

Then, inside the kernel, you could do:
```cpp
template <...>
__global__ void devKeyValSortColumnPerRow(...) {
  __shared__ SmemBlock<InType, BLOCK_SIZE, ITEMS_PER_THREAD>::TempStorage tempStorage;
....
}
```

Finally, to compute the smem size inside `sortColumnsPerRow`:
`size_t smemNeeded = sizeof(SmemBlock<InType, BLOCK_SIZE, ITEMS_PER_THREAD>::TempStorage);`

What say?

Again, I haven't compiled these code snippets. So, you might have to massage it accordingly!